simulator/infra/cache/LRUTagCache.h

```diff
@@ -0,0 +1,92 @@
+/**
+ * LRU cache module for BPU cache
+ * @author Denis Los
+*/
+
+#ifndef LRUTAGCACHE_H
+#define LRUTAGCACHE_H
+
+#include <cassert>
+
+#include <list>
+#include <unordered_map>
+#include <utility>
+
+// size_type should be a type of unsigned integer
+template <typename Key, typename size_type>
+class LRUTagCache
+{
+    public:
+        explicit LRUTagCache( size_type capacity)
+            : CAPACITY( capacity)
+        {
+            assert( capacity != 0u);
+
+            data.reserve( CAPACITY);
+            lru_hash.reserve( CAPACITY);
+        }
+
+        auto get_capacity() const { return CAPACITY; }
+        auto size() const { return number_of_elements; }
+        bool empty() const { return size() == 0u; }
+
+        auto find( const Key& key) const
+        {   
+            auto result = data.find( key);
+            return ( result != data.end()) 
+                     ? std::pair<bool, size_type>( true, result -> second)
+                     : std::pair<bool, size_type>( false, 0u);
+        }
+
+        void update( const Key& key)
+        {
+            auto data_it = data.find( key);
+            auto lru_it  = lru_hash.find( key);
+            assert( ( data_it == data.end()) == ( lru_it == lru_hash.end()));
+            
+            if ( data_it == data.end())
+                allocate( key);
+            else
+                lru_list.splice( lru_list.begin(), lru_list, lru_it->second);
+        }
+
+
+    private:
+        void allocate( const Key& key)
+        {
+            size_type value = 0;
+
+            if ( number_of_elements == CAPACITY)
+            {
+                // Delete least recently used element and save its value
+                const auto& lru_elem = lru_list.back();
+                lru_hash.erase( lru_elem);
+                value = data.find( lru_elem) -> second; 
+                data.erase( lru_elem);
+                lru_list.pop_back();
+            }
+            else 
+            {
+                value = number_of_elements;
+                number_of_elements++;
+            }
+
+            // Add a new element
+            data.emplace( key, value);
+            auto ptr = lru_list.insert( lru_list.begin(), key);
+            lru_hash.emplace( key, ptr);
+        }
+
+    private:
+        std::unordered_map<Key, size_type> data{};
+
+        std::list<Key> lru_list{};
+        std::unordered_map<Key, typename std::list<Key>::const_iterator> lru_hash{};
+
+        size_type number_of_elements = 0u;
+        size_type CAPACITY;
```

Comment:
Is there any use case to set "size_type" to anything other than "size_t"?

Reply:
It looks better for me not to convert size_t to uint32 and introduce size_type.

Reply:
I think uint32 should be used, as we will never have a cache with # of sets or ways exceeding the 32 bit.